### PR TITLE
Feature/us4 add contact

### LINKED ISSUE #4: Add a contact
## Value proposition

As a user
I want to add contacts to my contact list
So that I can manage new friends and family members

## Description

Image

## Acceptance criteria

- [x] Header is displaying a plus Icon on the right side of the logo
- [x] Clicking on the plus icon leads to a new page where I can add a new contact. 
- [x] "Add Contact Page" displays the contact header with the most important information (please see [#2](https://github.com/josch87/Gunther/issues/2)). Data is being displayed in real time when entered in controlled inputs
- [x] A form is displayed where you can enter the following information:
  - First name (mandatory)
  - Middle name
  - Last name
  - Nickname
  - Gender (Dropdown)
  - Birthday (date)
  - Instagram username
  - Twitter username
  - email & type
  - phone & type
  - address & type
  - Deceased ("true" if the contact passed away)
- [x] The type fields which are related to some contact fields are dropdowns which can have the following values: "Home", "Work" or "Other"
- [x] A Save and Cancel button are visible below the form
- [x] Clicking on the save button saves the entered form data and leads you to the detail page of the newly created contact
- [x] Clicking on the cancel button leads you back to the previous page you have been to.

### PR DESCRIPTION
**Pull request by josch87**
- closes #4

### CODE REVIEW

File: pages/_app.js [hunk 1]
```diff
@@ -2,17 +2,142 @@ import GlobalStyle from "../styles";
...
+    event.preventDefault();
+
+    const formData = new FormData(event.target);
+    const newContact = Object.fromEntries(formData);
+
+    const newContactId = uid();
+
+    const date = new Date();
+    const currentUtcDateTime = date.toISOString();
+
```
Comment (julialoeschel) [inline pages/_app.js:24]: all this form handling should be done in your form component. Try to keep the app.js as clean as possible.

File: pages/_app.js [hunk 2]
```diff
@@ -2,17 +2,142 @@ import GlobalStyle from "../styles";
...
+        platform: "Twitter",
+        username: newContact.twitter,
+      });
+    }
+    if (newContact.facebook) {
+      formattedContact.socialMedia.push({
+        platform: "Facebook",
+        username: newContact.facebook,
+      });
+    }
```
Comment (julialoeschel) [inline pages/_app.js:125]: all there if cases and pushing into an object, that you just created, seems a little too much to me. Try to find a shorter way of writing, check operators like `ternery`, `&&` or `||` to make your code better understandable. 

and: this sould be done in the Form component, not in app.js ;)

File: pages/_app.js [hunk 3]
```diff
@@ -2,17 +2,142 @@ import GlobalStyle from "../styles";
 import { useImmerLocalStorageState } from "@/hooks/useImmerLocalStorageState";
 import { contactsSampleData } from "@/data/SampleData";
 import Layout from "@/components/Layout/Layout";
+import { uid } from "uid";
+import { useRouter } from "next/router";
 
 export default function App({ Component, pageProps }) {
+  const router = useRouter();
   const [contacts, updateContacts] = useImmerLocalStorageState("contacts", {
```
Comment (julialoeschel) [inline pages/_app.js:10]: I am not sure why and how do you need UseImmer in your project, and if its the best option for the project in this moment. 
If you need to use Localstorage, we suggest the `use-local-storage-state` hook, if you need a state management, we suggest `zustand` (but I dont see that necessary at this point) 

please rethink if `Immer` is the best choice .

File: components/Button/Button.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+      {children}
+    </StyledButton>
+  );
+}
+
+export function PrimaryButton({ children, type, onClick }) {
+  return (
+    <StyledPrimaryButton type={type} onClick={onClick}>
+      {children}
+    </StyledPrimaryButton>
```
Comment (julialoeschel) [inline components/Button/Button.js:15]: this is basically the same component, with different styling. 
you could think about a porp, that tells if the Button is a primary or not and adust the styling accordingly.

**Review by julialoeschel — COMMENTED** [review]
Hey There, 
good job! 

I found some issues you should think about.
keep on going 🐜

File: pages/_app.js (hunk 1 above)
Comment (josch87) [reply pages/_app.js:24]: Moved almost all of handleAddNewContact to ContactForm.js "handleSubmitForm" function

File: pages/_app.js (hunk 2 above)
Comment (josch87) [reply pages/_app.js]: @julialoeschel 

GPT made it easier for me using .filter at the end of the array. For example: 

```
      email: [
        {
          value: newContact.emailOne,
          type: newContact.emailOneType,
        },
        {
          value: newContact.emailTwo,
          type: newContact.emailTwoType,
        },
      ].filter((email) => email.value),
      phone: [
        {
          value: newContact.phoneOne,
          type: newContact.phoneOneType,
        },
        {
          value: newContact.phoneTwo,
          type: newContact.phoneTwoType,
        },
      ].filter((phone) => phone.value),
      address: ...
```

How can I do the same for socialMedia array?
```
    if (newContact.instagram) {
      formattedContact.socialMedia.push({
        platform: "Instagram",
        username: newContact.instagram,
      });
    }
    if (newContact.twitter) {
      formattedContact.socialMedia.push({
        platform: "Twitter",
        username: newContact.twitter,
      });
    }
    if (newContact.facebook) {
      formattedContact.socialMedia.push({
        platform: "Facebook",
        username: newContact.facebook,
      });
    }
```

I tried with a ternary but how do I not return nothing when the condition is falsy? 
```
      socialMedia: [
        newContact.instagram
          ? {
              platform: "Instagram",
              username: newContact.instagram,
            }
          : null,
      ],
```
Trying the above I get NULL into my array which leads to an error when displaying my data later on.
Screenshot 2023-07-13 at 16 45 23

File: pages/_app.js (hunk 3 above)
Comment (josch87) [reply pages/_app.js:10]: I removed the hook useImmerLocalStorageState and used useLocalStorageState instead

File: components/Button/Button.js (hunk 4 above)
Comment (josch87) [reply components/Button/Button.js:15]: I changed it to be just one component with an if condition and a prop called "buttonType". 

```
export default function Button({ children, type, onClick, buttonType }) {
  if (buttonType === "primary") {
    return (
      <StyledPrimaryButton type={type} onClick={onClick}>
        {children}
      </StyledPrimaryButton>
    );
  }

  return (
    <StyledButton type={type} onClick={onClick}>
      {children}
    </StyledButton>
  );
}
```

@julialoeschel: Is this what you were thinking of?

**Review by ryliecc — APPROVED** [review]
LGTM, easy to understand and I think you successfully implemented all of Julia's suggestions :)

File: pages/_app.js (hunk 2 above)
Comment (josch87) [reply pages/_app.js]: After speaking with Jessica and Dominik it works just the same:

```
      socialMedia: [
        {
          platform: "Instagram",
          username: newContact.instagram,
        },
        {
          platform: "Twitter",
          username: newContact.twitter,
        },
        {
          platform: "Facebook",
          username: newContact.facebook,
        },
      ].filter((socialMedia) => socialMedia.username),
```

Comment (josch87) [reply pages/_app.js]:
```
.filter((socialMedia) => socialMedia.username)
```
is like saying
```
.filter((socialMedia) => (socialMedia.username ? true : false))
```

**Pull request merged**